### PR TITLE
TCP connection support

### PR DESCRIPTION
**Pull request by DemiMarie**
Only IP addresses are supported.  This is based on #139.

### CODE REVIEW

**Comment by marmarek** [comment]
Test test_connect_socket_tcp_ipv6 fail:
```
family = <AddressFamily.AF_INET6: 10>, service = 'qubes.SocketService+++1+65532'
host = '::1', port = 65532, accept = True
    def _test_tcp_raw(self, family: int, service: str, host: str, port: int, accept=True):
        server = socket.socket(family, socket.SOCK_STREAM, socket.IPPROTO_TCP)
        server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
>       server.bind((host, port))
E       OSError: [Errno 99] Cannot assign requested address
```

**Comment by DemiMarie** [comment]
> Test test_connect_socket_tcp_ipv6 fail:
> 
> ```
> family = <AddressFamily.AF_INET6: 10>, service = 'qubes.SocketService+++1+65532'
> host = '::1', port = 65532, accept = True
>     def _test_tcp_raw(self, family: int, service: str, host: str, port: int, accept=True):
>         server = socket.socket(family, socket.SOCK_STREAM, socket.IPPROTO_TCP)
>         server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
> >       server.bind((host, port))
> E       OSError: [Errno 99] Cannot assign requested address
> ```

Is CI misconfigured?  The test definitely passes locally.

**Comment by DemiMarie** [comment]
Of course, I can skip this test, but I would prefer to fix this in CI.

**Comment by DemiMarie** [comment]
> Test test_connect_socket_tcp_ipv6 fail:
> 
> ```
> family = <AddressFamily.AF_INET6: 10>, service = 'qubes.SocketService+++1+65532'
> host = '::1', port = 65532, accept = True
>     def _test_tcp_raw(self, family: int, service: str, host: str, port: int, accept=True):
>         server = socket.socket(family, socket.SOCK_STREAM, socket.IPPROTO_TCP)
>         server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
> >       server.bind((host, port))
> E       OSError: [Errno 99] Cannot assign requested address
> ```

This is probably a Docker bug: <https://github.com/moby/moby/issues/33099>.  I suggest either switching to system-wide Podman, or upgrading to a version of Docker that isn’t affected (assuming such a version exists).

**Comment by marmarek** [comment]
Now this needs a rebase...

**Comment by DemiMarie** [comment]
> Now this needs a rebase...

fixed

**Comment by marmarek** [comment]
testing didn't went too far, failed already at qubes-dom0-update (after installing updated qrexec):
```
2024-04-18 07:57:16.171 qrexec-client[4351]: exec.c:601:find_qrexec_service: Service not found: qubes.ReceiveUpdates+
2024-04-18 07:57:16.235 qrexec-client[4351]: qrexec-daemon-common.c:295:handle_failed_exec: failed to spawn process, exiting
2024-04-18 07:57:16.236 qrexec-daemon[2395]: qrexec-daemon.c:733:reap_children: qrexec-policy-exec for connection SOCKET14 exited with code 127, but the response (allow) was already sent
```

`/etc/qubes-rpc/qubes.ReceiveUpdates` is a symlink to executable.

**Comment by DemiMarie** [comment]
> testing didn't went too far, failed already at qubes-dom0-update (after installing updated qrexec):
> 
> ```
> 2024-04-18 07:57:16.171 qrexec-client[4351]: exec.c:601:find_qrexec_service: Service not found: qubes.ReceiveUpdates+
> 2024-04-18 07:57:16.235 qrexec-client[4351]: qrexec-daemon-common.c:295:handle_failed_exec: failed to spawn process, exiting
> 2024-04-18 07:57:16.236 qrexec-daemon[2395]: qrexec-daemon.c:733:reap_children: qrexec-policy-exec for connection SOCKET14 exited with code 127, but the response (allow) was already sent
> ```
> 
> `/etc/qubes-rpc/qubes.ReceiveUpdates` is a symlink to executable.

Whoops, that means there is a missing unit test.  Will fix.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: qrexec/tests/socket/agent.py [hunk 1]
```diff
@@ -686,6 +703,201 @@ def test_connect_socket_no_metadata(self):
...
+    def test_connect_socket_tcp_missing_host(self):
+        """
+        /dev/tcp socket service, neither host nor port provided.
+        """
+        socket_path = os.path.join(
+            self.tempdir, "rpc", "qubes.SocketService"
+        )
+        path = "/dev/tcp"
+        os.symlink(path, socket_path)
+        target, dom0 = self.execute_qubesrpc("qubes.SocketService+127.0.0.1", "domX")
```
Comment (marmarek) [inline qrexec/tests/socket/agent.py:824]:
```suggestion
        target, dom0 = self.execute_qubesrpc("qubes.SocketService+", "domX")
```
?

**Review by marmarek — CHANGES_REQUESTED** [review]

**Review by marmarek — APPROVED** [review]

**Pull request merged**